core/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java

```diff
@@ -43,32 +43,127 @@
+ *
+ * We don't always know if this task will be a parent or child task when it's created, because if slices is set to "auto" it may
+ * be either depending on the number of shards in the source indices. We figure that out when the request is handled and set it on this
+ * class with {@link #setSliceChildren(int)} or {@link #setSliceChild(float, Integer)}.
  */
-public abstract class BulkByScrollTask extends CancellableTask {
+public class BulkByScrollTask extends CancellableTask {
+
+    private ParentBulkByScrollWorker parentWorker;
+    private ChildBulkByScrollWorker childWorker;
+
     public BulkByScrollTask(long id, String type, String action, String description, TaskId parentTaskId) {
         super(id, type, action, description, parentTaskId);
     }
 
+    @Override
+    public BulkByScrollTask.Status getStatus() {
+        if (isSlicesParent()) {
+            return parentWorker.getStatus();
+        }
+
+        if (isSliceChild()) {
+            return childWorker.getStatus();
+        }
+
+        return emptyStatus();
+    }
+
+    private BulkByScrollTask.Status emptyStatus() {
+        return new Status(Collections.emptyList(), getReasonCancelled());
+    }
+
     /**
-     * The number of sub-slices that are still running. {@link WorkingBulkByScrollTask} will always have 0 and
-     * {@link ParentBulkByScrollTask} will return the number of waiting tasks. Used to decide how to perform rethrottling.
+     * Returns true if this task is a parent of other sliced tasks. False otherwise
      */
-    public abstract int runningSliceSubTasks();
+    public boolean isSlicesParent() {
+        return parentWorker != null;
+    }
 
     /**
-     * Apply the {@code newRequestsPerSecond}.
+     * Sets this task to be a parent task for {@code slices} sliced subtasks
      */
-    public abstract void rethrottle(float newRequestsPerSecond);
+    public void setSliceChildren(int slices) {
+        if (isSlicesParent()) {
+            throw new IllegalStateException("Parent worker is already set");
+        }
+        if (isSliceChild()) {
+            throw new IllegalStateException("Worker is already set as child");
+        }
 
-    /*
-     * Overridden to force children to return compatible status.
+        parentWorker = new ParentBulkByScrollWorker(this, slices);
+    }
+
+    /**
+     * Returns the worker object that tracks the state of sliced subtasks. Throws IllegalStateException if this task is not set to be
+     * a parent task.
      */
-    public abstract BulkByScrollTask.Status getStatus();
+    public ParentBulkByScrollWorker getSlicesParentWorker() {
+        if (!isSlicesParent()) {
+            throw new IllegalStateException("This task is not set as a parent worker");
+        }
+        return parentWorker;
+    }
 
     /**
-     * Build the status for this task given a snapshot of the information of running slices.
+     * Returns true if this task is a child task that performs search requests. False otherwise
```

Comment:
It'll return true for single sliced requests since they get set with a child worker, even if `slices` was `auto`. I agree the names aren't great and there's a lot of overlap, I'll see if I can find some better ones